File: lib/embed-fonts.py
"""
Voegt Tungsten font embedding toe aan een bestaande PPTX.
Font embedding in OOXML werkt via:
- ppt/fonts/ map met de font binaries
- <p:embeddedFontLst> in presentation.xml met verwijzingen
- Relationships in presentation.xml.rels
"""
import zipfile, shutil, os, sys
from fontTools.ttLib import TTFont
from lxml import etree

FONTS = {
    'Tungsten Bold': { 'regular': 'public/fonts/Tungsten-Bold.otf' },
    'Tungsten Book': { 'regular': 'public/fonts/Tungsten-Book.otf' },
    'Tungsten SemiBold': { 'regular': 'public/fonts/Tungsten-Semibold.otf' },
}

def otf_to_embedding_bytes(otf_path):
    with open(otf_path, 'rb') as f:
        return f.read()

def embed_fonts_in_pptx(input_pptx, output_pptx, fonts):
    shutil.copy(input_pptx, output_pptx)

    P_NS   = 'http://schemas.openxmlformats.org/presentationml/2006/main'
    A_NS   = 'http://schemas.openxmlformats.org/drawingml/2006/main'
    REL_NS = 'http://schemas.openxmlformats.org/officeDocument/2006/relationships'
    CT_NS  = 'http://schemas.openxmlformats.org/package/2006/content-types'
    FONT_REL_TYPE = 'http://schemas.openxmlformats.org/officeDocument/2006/relationships/font'
    CT_FONT = 'application/x-fontdata'

    with zipfile.ZipFile(output_pptx, 'r') as zin:
        files = {name: zin.read(name) for name in zin.namelist()}

    prs_tree  = etree.fromstring(files['ppt/presentation.xml'])
    rels_tree = etree.fromstring(files['ppt/_rels/presentation.xml.rels'])
    ct_tree   = etree.fromstring(files['[Content_Types].xml'])

    existing_defaults = {e.get('Extension') for e in ct_tree.findall(f'{{{CT_NS}}}Default')}

    font_lst = prs_tree.find(f'{{{P_NS}}}embeddedFontLst')
    if font_lst is None:
        font_lst = etree.SubElement(prs_tree, f'{{{P_NS}}}embeddedFontLst')

    rel_id_counter = 100

    variant_tags = {
        'regular':    'regular',
        'bold':       'bold',
        'italic':     'italic',
        'boldItalic': 'boldItalic',
    }

    for font_name, variants in fonts.items():
        font_el  = etree.SubElement(font_lst, f'{{{P_NS}}}embeddedFont')
        font_def = etree.SubElement(font_el, f'{{{A_NS}}}font')
        font_def.set('typeface', font_name)
        font_def.set('charset', '0')

        for variant_key, otf_path in variants.items():
            if not os.path.exists(otf_path):
                print(f'⚠️  Font niet gevonden: {otf_path}', file=sys.stderr)
                continue

            rel_id = f'rFontId{rel_id_counter}'
            rel_id_counter += 1

            font_bytes    = otf_to_embedding_bytes(otf_path)
            font_filename = f'font{rel_id_counter}.odttf'
            font_path_in_zip = f'ppt/fonts/{font_filename}'
            files[font_path_in_zip] = font_bytes

            rel_el = etree.SubElement(rels_tree, 'Relationship')
            rel_el.set('Id',     rel_id)
            rel_el.set('Type',   FONT_REL_TYPE)
            rel_el.set('Target', f'fonts/{font_filename}')

            tag        = variant_tags.get(variant_key, 'regular')
            variant_el = etree.SubElement(font_el, f'{{{P_NS}}}{tag}')
            variant_el.set(f'{{{REL_NS}}}id', rel_id)

            if 'odttf' not in existing_defaults:
                ct_el = etree.SubElement(ct_tree, f'{{{CT_NS}}}Default')
                ct_el.set('Extension',   'odttf')
                ct_el.set('ContentType', CT_FONT)
                existing_defaults.add('odttf')

    files['ppt/presentation.xml'] = etree.tostring(
        prs_tree, xml_declaration=True, encoding='UTF-8', standalone=True)
    files['ppt/_rels/presentation.xml.rels'] = etree.tostring(
        rels_tree, xml_declaration=True, encoding='UTF-8', standalone=True)
    files['[Content_Types].xml'] = etree.tostring(
        ct_tree, xml_declaration=True, encoding='UTF-8', standalone=True)

    with zipfile.ZipFile(output_pptx, 'w', zipfile.ZIP_DEFLATED) as zout:
        for name, data in files.items():
            zout.writestr(name, data)

    print(f'✅ Fonts embedded: {output_pptx}')
    return output_pptx


if __name__ == '__main__':
    if len(sys.argv) < 3:
        print('Gebruik: python3 lib/embed-fonts.py <input.pptx> <output.pptx>', file=sys.stderr)
        sys.exit(1)
    embed_fonts_in_pptx(sys.argv[1], sys.argv[2], FONTS)

File: lib/embed-fonts.js
// lib/embed-fonts.js
// Pure Node.js font embedding voor PPTX via JSZip + @xmldom/xmldom.
// Voegt Tungsten font-binaries toe aan de OOXML-structuur zodat
// PowerPoint de fonts meeneemt ook als ze niet op de doelmachine staan.

import JSZip from 'jszip';
import { DOMParser, XMLSerializer } from '@xmldom/xmldom';
import { readFile } from 'node:fs/promises';
import path from 'node:path';

const P_NS   = 'http://schemas.openxmlformats.org/presentationml/2006/main';
const A_NS   = 'http://schemas.openxmlformats.org/drawingml/2006/main';
const R_NS   = 'http://schemas.openxmlformats.org/officeDocument/2006/relationships';
const REL_NS = 'http://schemas.openxmlformats.org/package/2006/relationships';
const CT_NS  = 'http://schemas.openxmlformats.org/package/2006/content-types';
const FONT_REL_TYPE = 'http://schemas.openxmlformats.org/officeDocument/2006/relationships/font';
const CT_FONT = 'application/x-fontdata';

const FONTS = [
  { name: 'Tungsten Bold',     file: 'Tungsten-Bold.otf',     variant: 'regular' },
  { name: 'Tungsten Book',     file: 'Tungsten-Book.otf',     variant: 'regular' },
  { name: 'Tungsten SemiBold', file: 'Tungsten-Semibold.otf', variant: 'regular' },
];

const parser     = new DOMParser();
const serializer = new XMLSerializer();

function parseXml(str) {
  return parser.parseFromString(str, 'text/xml');
}

function serializeXml(doc) {
  return serializer.serializeToString(doc);
}

export async function embedFonts(pptxBuffer) {
  const zip = await JSZip.loadAsync(pptxBuffer);

  // ── Lees bestaande XML-bestanden ────────────────────────────────────────────
  const prsXml   = await zip.file('ppt/presentation.xml').async('string');
  const relsXml  = await zip.file('ppt/_rels/presentation.xml.rels').async('string');
  const ctXml    = await zip.file('[Content_Types].xml').async('string');

  const prsDoc   = parseXml(prsXml);
  const relsDoc  = parseXml(relsXml);
  const ctDoc    = parseXml(ctXml);

  const prsRoot  = prsDoc.documentElement;
  const relsRoot = relsDoc.documentElement;
  const ctRoot   = ctDoc.documentElement;

  // ── [Content_Types].xml: voeg odttf toe als die er nog niet is ───────────
  const existingDefaults = Array.from(ctRoot.getElementsByTagNameNS(CT_NS, 'Default'))
    .map(el => el.getAttribute('Extension'));
  if (!existingDefaults.includes('odttf')) {
    const defaultEl = ctDoc.createElementNS(CT_NS, 'Default');
    defaultEl.setAttribute('Extension', 'odttf');
    defaultEl.setAttribute('ContentType', CT_FONT);
    ctRoot.appendChild(defaultEl);
  }

  // ── Bestaande rel-ID's ophalen om conflicten te vermijden ────────────────
  const existingIds = new Set(
    Array.from(relsRoot.getElementsByTagNameNS(REL_NS, 'Relationship'))
      .map(el => el.getAttribute('Id'))
  );
  let relCounter = 200;
  function nextRelId() {
    while (existingIds.has(`rFontId${relCounter}`)) relCounter++;
    const id = `rFontId${relCounter}`;
    existingIds.add(id);
    relCounter++;
    return id;
  }

  // ── embeddedFontLst: zoek of maak aan ────────────────────────────────────
  let fontLst = prsRoot.getElementsByTagNameNS(P_NS, 'embeddedFontLst')[0];
  if (!fontLst) {
    fontLst = prsDoc.createElementNS(P_NS, 'p:embeddedFontLst');
    prsRoot.appendChild(fontLst);
  }

  // ── Verwerk elk font ──────────────────────────────────────────────────────
  const fontsDir = path.resolve(process.cwd(), 'public/fonts');

  for (const { name, file, variant } of FONTS) {
    const fontPath = path.join(fontsDir, file);
    let fontBytes;
    try {
      fontBytes = await readFile(fontPath);
    } catch {
      console.warn(`[PPTX] font niet gevonden, overgeslagen: ${fontPath}`);
      continue;
    }

    const relId        = nextRelId();
    const fontFilename = `${relId}.odttf`;
    const zipPath      = `ppt/fonts/${fontFilename}`;

    // Voeg font binary toe aan ZIP
    zip.file(zipPath, fontBytes);

    // Relationship in presentation.xml.rels
    const relEl = relsDoc.createElementNS(REL_NS, 'Relationship');
    relEl.setAttribute('Id',     relId);
    relEl.setAttribute('Type',   FONT_REL_TYPE);
    relEl.setAttribute('Target', `fonts/${fontFilename}`);
    relsRoot.appendChild(relEl);

    // <p:embeddedFont> in presentation.xml
    const fontEl  = prsDoc.createElementNS(P_NS, 'p:embeddedFont');
    const fontDef = prsDoc.createElementNS(A_NS, 'a:font');
    fontDef.setAttribute('typeface', name);
    fontDef.setAttribute('charset',  '0');
    fontEl.appendChild(fontDef);

    const variantEl = prsDoc.createElementNS(P_NS, `p:${variant}`);
    variantEl.setAttributeNS(R_NS, 'r:id', relId);
    fontEl.appendChild(variantEl);

    fontLst.appendChild(fontEl);
  }

  // ── Schrijf gewijzigde XML terug ─────────────────────────────────────────
  zip.file('ppt/presentation.xml',            serializeXml(prsDoc));
  zip.file('ppt/_rels/presentation.xml.rels', serializeXml(relsDoc));
  zip.file('[Content_Types].xml',             serializeXml(ctDoc));

  return zip.generateAsync({ type: 'nodebuffer', compression: 'DEFLATE' });
}
